Update contributing guidelines (#740)
cometbft/cometbft: CONTRIBUTING.md

develop a BFT consensus engine robust enough to support permissionless
value-carrying networks. While all contributions are welcome, contributors
should bear this goal in mind in deciding if they should target the main
CometBFT project or a potential fork.

## Overview

When targeting the main CometBFT project, following the processes outlined in
this document will lead to the best chance of landing changes in a release.

### Core team responsibility

> First and foremost, the CometBFT core team is responsible for maintaining this

Reviewer: This might be simpler and more accurate:

```suggestion
The CometBFT core team is responsible for stewarding this
```

project over time. This means that the core team needs to understand the nature
of, and agree to maintain, all of the changes that land on `main` or a backport

> branch. It may cost you a few days/weeks' worth of your time to submit a
> particular change, but it may cost the core team far more to maintain that
> change over the years.

Reviewer: We could rephrase this:

```suggestion
branch. It may cost a few days/weeks' worth of time to _submit_ a
particular change, but _maintaining_ that change over the years has a
much higher cost that the core team will bear.
```

security fixes), but this is rare.

Each stage of the process is aimed at creating feedback cycles which align
contributors and maintainers to make sure:

- Contributors don’t waste their time implementing/proposing features which
  won't land in `main`.
- Maintainers have the necessary context in order to support and review
  contributions.

```mermaid
flowchart LR
    complexity{Problem\ncomplexity}
    issue("New issue\n(Problem articulation\nfor discussion)")
    clarity{"Problem +\nsolution clarity"}
    rfc("RFC pull request(s)")
    rfc_merge("Merge RFC to main")
    risk{"Solution\ncomplexity/risk"}
    adr("ADR + PoC\npull request(s)")
    adr_merge("Merge ADR to main\nand create tracking issue")
    pr("Solution\npull request(s)")
    merge("Merge to main/backport\nor feature branch")

    complexity --"Low/Moderate/High"--> issue
    complexity --Trivial--> pr
    issue --> clarity
    clarity --High--> risk
    clarity --Low--> rfc
    rfc --Approved--> rfc_merge
    risk --"Moderate/High"--> adr
    adr --"ADR accepted by core team"--> adr_merge
    adr_merge --> pr
    risk --Low--> pr
    pr --Approved--> merge
```

> ### GitHub issues

Reviewer: 💯
